crypto/rand/rand_lib.c

```diff
@@ -178,6 +178,8 @@ size_t rand_drbg_get_entropy(RAND_DRBG *drbg,
                                    prediction_resistance,
                                    NULL, 0) != 0)
                 bytes = bytes_needed;
+            drbg->reseed_next_counter
+                = tsan_load(&drbg->parent->reseed_prop_counter);
```

Comment:
Yes, this is likely not needed, but I suspected hat tsan would complain if it is sometimes
read as atomic and sometimes as non-atomic.

Reply:
I meant the entire assignment to `reseed_next_counter` not just the `tsan_load`.  Everything time `reseed_next_counter` is stored to, it is read from `reseed_prop_counter`.  Soon after, it is stored back to `reseed_prop_counter`.  I.e. it looks like it could be made into a local in the functions that do this rather than being carried in the structure.

Reply:
Unless I'm missing something ??

Reply:
The value is consistent with the quality of the acquired entropy if it is read before the
rand_drbg_unlock below.

Reply:
I'm not explaining my concern well enough.  Look at all the uses of `reseed_next_counter` throughout the code.  At no point is it read from except soon after it is assigned to.  I.e. it doesn't need to be persistent.  Both `RAND_DRBG_instantiate` and `RAND_DRBG_reseed` take the form:

```
drbg->reseed_next_counter = drbg->reseed_prop_counter + 1;
...
drbg->reseed_prop_counter = drbg->reseed_next_counter;
```

`reseed_next_counter` is not used anywhere else, except for this assignment.


Would it make sense to change the code in each function to:

```
unsigned int reseed_next_counter;
...
reseed_next_counter = drbg->reseed_prop_counter + 1;
...
drbg->reseed_prop_counter = reseed_next_counter;
```

and to remove `reseed_next_counter` from the structure?

Reply:
No. The code looks like this:

```
drbg->reseed_next_counter = drbg->reseed_prop_counter + 1;
drbg->get_entropy(drbg, ...);
drbg->reseed_prop_counter = drbg->reseed_next_counter;
```
get_entropy may or may not change reseed_next_counter to the parent's
reseed_prop_counter.
If get_entropy is overwritten, it will be impossible to access reseed_next_counter,
as well as drbg->parent, drbg->pool etc.
It would normally be easy to add a new out-parameter to get_entropy but unfortunately
we have to go to extremes, to keep the get_entropy function signature interface as is,
which is part of the public API, while we may change opaque structures like drbg
if we have to.

Reply:
+1: Since we kept the DRBG chaining private, it makes sense not to expose the reseed_next_counter either.

Reply:
Thanks for the explanation, it makes sense now.